src/passes/ConstantFieldPropagation.cpp

```diff
@@ -176,48 +238,72 @@ struct FunctionOptimizer : public WalkerPass<PostWalker<FunctionOptimizer>> {
     // constant value. (Leave it to further optimizations to get rid of the
     // ref.)
     auto* value = makeExpression(info, heapType, curr);
-    auto* replacement = builder.blockify(
-      builder.makeDrop(builder.makeRefAs(RefAsNonNull, curr->ref)));
-    // If this get is sequentially consistent, then it synchronizes with other
-    // threads at least by participating in the global order of sequentially
-    // consistent operations. Preserve that effect by replacing the access with
-    // a fence.
-    assert(curr->order != MemoryOrder::AcqRel);
-    if (curr->order == MemoryOrder::SeqCst) {
-      replacement = builder.blockify(replacement, builder.makeAtomicFence());
+    auto* replacement = makeDroppedRef(curr);
+    replacement = maybeAddFence(replacement, curr->order);
+    replaceCurrent(builder.blockify(replacement, value));
+    changed = true;
+  }
+
+  template<typename T>
+  std::optional<std::pair<HeapType, PossibleConstantValues>>
+  shouldOptimizeRMW(T* curr) {
+    HeapType heapType;
+    if (auto type = getRelevantHeapType(curr)) {
+      heapType = *type;
+    } else {
+      return std::nullopt;
+    }
+
+    // Get the info about the field. Since RMWs can only copy or mutate the
+    // value, we always have something recorded.
+    PossibleConstantValues info = getInfo(heapType, curr->index);
+    assert(info.hasNoted() && "unexpected lack of info for RMW");
+
+    if (curr->order == MemoryOrder::AcqRel) {
+      // See comment on visitStructGet for why we don't optimize here.
+      return std::nullopt;
     }
+
+    if (!info.isConstant()) {
+      // Optimizing using ref.test is not an option here because that only works
+      // on immutable fields, but RMW operations always access mutable fields.
+      return std::nullopt;
+    }
+
+    // We can optimize.
+    return std::pair(heapType, info);
+  }
+
+  void visitStructRMW(StructRMW* curr) {
+    auto typeAndInfo = shouldOptimizeRMW(curr);
+    if (!typeAndInfo) {
+      return;
+    }
+    auto& [type, info] = *typeAndInfo;
+    Builder builder(*getModule());
+    auto* value = makeExpression(info, type, curr);
```

Comment:
Should we assert that the RMW op is xchg? and/or/etc operations do not allow the value to be constant/optimizable.